XenModel/PoolJoinRules.cs

```diff
@@ -461,6 +467,11 @@ private static bool LicenseMismatch(Host slave, Host master)
             return slaveEdition != Host.Edition.Free && masterEdition != Host.Edition.Free && slaveEdition != masterEdition;
         }
 
+        private static bool UnlicensedMasterMaxNumberHostReached(IXenConnection connection, Host master)
+        {
+            return connection.Cache.HostCount > 2 && master.IsFreeLicense();

```

Comment:
This check only needs to be performed if the pool has the pool size restriction (the "restrict_pool_size" license key exists and it is true).
Then `master.IsFreeLicense()` is not needed